src/viewer.coffee

```diff
@@ -0,0 +1,156 @@
+#
+#   You may provide...
+#     onRender: [function]
+#       Callback to be run after each diagram render operation, with `this`
+#       taking the value of the ApparatusViewer.
+#
+# You can read and write the Apparatus project displayed in `myViewer` using
+# `myViewer.project`.
+#
+# As a convenience, `myViewer.getAttributeByLabel(label)` will grab the
+# attribute node in the displayed element with label `label`, provided this
+# label is unique.
+
+window.ApparatusViewer = class ApparatusViewer
+  constructor: (@options) ->
+    # TODO: Add selection of which element to view?
+    @_getElement()
+    @_getProjectAndLoad()
+
+  _load: ->
+    # Now we can actually load the diagram!
+
+    # Use regionOfInterest to compute view matrix:
+    {regionOfInterest} = @options
+    rect = @element.getBoundingClientRect()
+    scaleFactor = Math.min(
+      rect.width / (regionOfInterest.x[1] - regionOfInterest.x[0]),
+      rect.height / (regionOfInterest.y[1] - regionOfInterest.y[0]))
+    matrix = new Util.Matrix()
+    matrix = matrix.scale(scaleFactor, scaleFactor)
+    matrix = matrix.translate(
+      -(regionOfInterest.x[1] + regionOfInterest.x[0])/2,
+      -(regionOfInterest.y[1] + regionOfInterest.y[0])/2)
+    @project.editingElement.viewMatrix = matrix
+
+    @_render()
+
+    @willRefreshNextFrame = false
+    for eventName in refreshEventNames
+      window.addEventListener(eventName, => @_refresh())
+
+  _getElement: ->
+    if @options.element and @options.selector
+      throw "Do not provide both `element` and `selector` as options"
+
+    if @options.element?
+      @element = @options.element
+    else if @options.selector?
+      @element = document.querySelector(@options.selector)
+    else
+      throw "Either `element` or `selector` must be provided as an option"
+
+  _getProjectAndLoad: ->
+    if @options.projectData and @options.url
+      throw "Do not provide both `projectData` and `url`"
+
+    if @options.projectData?
+      @projectData = @options.projectData
+      @_deserializeProjectData()
+      @_load()
+    else if @options.url?
+      xhr = new XMLHttpRequest()
+      xhr.onreadystatechange = (e) =>
+        if xhr.readyState == 4 and xhr.status == 200
+          @projectData = JSON.parse(xhr.responseText)
+          @_deserializeProjectData()
+          @_load()
+      xhr.open("GET", @options.url)
+      xhr.send()
+    else
+      throw "Either `projectData` or `url` must be provided as an option"
+
+  _deserializeProjectData: ->
+    serializer = Model.SerializerWithBuiltIns.getSerializer()
+    @project = serializer.dejsonify(@projectData)
+
+  _render: ->
+    Dataflow.run =>
+      R.render(R.Viewer({@project}), @element)
+    @options.onRender?.apply(this)
+
+  _refresh: ->
+    return if @willRefreshNextFrame
+    @willRefreshNextFrame = true
+    requestAnimationFrame =>
+      @_render()
+      @willRefreshNextFrame = false
+
+  getAttributeByLabel: (label) ->
```

Comment:
I'm not sure if exposing the underlying model is the best idea in the long run. A more robust option might be for the viewer to have its own documented API. You've already started writing this documentation at the start of this file, explaining how to construct a viewer.

The API would have hooks for common use cases: listening for render events, getting attribute values (by their label), setting attribute values and/or expressions, perhaps switching to a different "scene" (by changing editingElement), etc. It could also provide a hook into the underlying model for people who want to do more advanced things, but the common things would have a robust API that we'd support even if we evolved the underlying model. Happy to brainstorm more on this...